### PR TITLE
Travis CI support

### PR DESCRIPTION
**Pull request by godbyk**
This allows the PDF to be built using [Travis CI](https://travis-ci.org/godbyk/draft). Handy for catching syntax errors in pull requests.

The build script could be expanded to add other checks in the future, too (for example, raise errors on too overfull lines, missing cross-reference labels).

### CODE REVIEW

**Comment by tkoeppe** [comment]
Please squash.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    - ppa:texlive-backports/ppa
+    packages:
+    - latexmk
+    - poppler-utils
+    - texlive-binaries
+    - texlive-fonts-recommended
+    - texlive-latex-base
+    - texlive-latex-extra
+    - texlive-latex-recommended
+    - texlive-xetex
```
Comment (tkoeppe) [inline .travis.yml]: I don't think we need xetex.

**Comment by godbyk** [comment]
@tkoeppe I've squashed the commits. I also removed the unnecessary package and unused repository. Thanks!

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+script:
+  - pushd source
+  - make -j2
```
Comment (godbyk) [inline .travis.yml]: @tkoeppe Only because I know `make` works and have never tried building it with `latexmk`. If `latexmk` works and is preferred, I'm happy to use it instead.

Comment (tkoeppe) [inline .travis.yml]: Well, either use `make` and don't install `latexmk`, or use `latexmk`, I'd say...

Comment (godbyk) [inline .travis.yml]: @tkoeppe I'm not picky. Which method is preferred?

Comment (tkoeppe) [inline .travis.yml]: `latexmk`

Comment (godbyk) [inline .travis.yml]: @tkoeppe I'll give `latexmk` a try. I think I'll have to duplicate some of the effort from the `Makefile` to the `.travis.xml` file (e.g., building the diagram PDFs with `dot`, generating the grammar and cross-references appendices) since `latexmk` won't do that for us.

Is it worth the effort or is it better to rely on the `Makefile`?

Comment (tkoeppe) [inline .travis.yml]: Ah, it maybe that `latexmk` doesn't go the whole way and is only useful for the common case of small changes. In that case stick with `make`, sure.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  - popd
+
+sudo: false
+
+addons:
+  apt:
+    packages:
+    - latexmk
```
Comment (tkoeppe) [inline .travis.yml]: So, why are you installing `latexmk` then?

Comment (godbyk) [inline .travis.yml]: Because I started by copying the list of required packages from the README.

We could set up a build matrix and have it build using `latexmk` and with the `Makefile` to exercise both build options, too, I s'pose.

Installing the `latexmk` package isn't terribly time-consuming or onerous in any case, though.

File: .travis.yml (hunk 2 above)
Comment (tkoeppe) [inline .travis.yml]: On second thought, please do use `latexmk` for the test. The regeneration of the xrefs and grammar and plots is not a transient operation; and thus `make` is a tool for the author, not the user (despite what we say in the readme). If `make` were a user-facing tool, then we wouldn't have the grammar and xrefs checked in.

Note also that just `make` doesn't regenerate anything, since it only runs `rebuild` by default.

In summary, the semantics of `latexmk -pdf std` seem to be exactly what we want.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    - poppler-utils
+    - texlive-binaries
+    - texlive-fonts-recommended
+    - texlive-latex-base
+    - texlive-latex-extra
+    - texlive-latex-recommended
+    - texlive-binaries
+    - graphviz
+    - lmodern
+
```
Comment (tkoeppe) [inline .travis.yml]: Gratuitous newline, remove.

**Comment by tkoeppe** [comment]
@zygoloid: What do you think of adding Travis testing? I think it may require you to link the repository to Travis somehow and give them access. Running `latexmk` seems like a decent smoke test, though on the other hand I don't know how much lifting power that would have.

**Comment by zygoloid** [comment]
Sounds like a good idea to me. What sort of access do they require? Presumably push access isn't required for them to do this integration?

In addition to latexmk, it might be interesting to run the various regeneration steps (makegram etc) and check that the checked-in outputs of these steps don't change (though I do wonder how much sense it makes to check in a generated file in the first place...).

**Comment by tkoeppe** [comment]
Setup instructions here: https://docs.travis-ci.com/user/getting-started/#To-get-started-with-Travis-CI%3A

I agree in principle that checking in generated files is iffy, but at least the current setup separates LaTeX editing from our own homebrew scripts. I wouldn't want to *mandate* that any user be able to execute our scripts. However, I would be quite happy to have a test that a simple LaTeX edit does't *change* the output of the generated files, as you suggested, so it would indeed be nice if we could make Travis run that test.

**Comment by godbyk** [comment]
I would recommend setting up a matrix and having Travis CI build the files with both latexmk and using the build scripts.

**Comment by godbyk** [comment]
I've rebased this PR against master and updated the `.travis.yml` file to do the following:

Travis CI will build `std.pdf` in four different ways:

  1. Using `latexmk`.
  2. Using `make`.
  3. Manually per the instructions in `README.rst`.
  4. Manually (same as 3), but also regenerating the grammar annex, cross-references, and figures.

It also regenerates the grammar annex, cross-references, and figures and compares them to the pre-generated versions checked into the repository.

You can view the results of this PR at <https://travis-ci.org/godbyk/draft/builds/137449847>.

File: .travis.yml (hunk 4 above)
Comment (tkoeppe) [inline .travis.yml]: @godbyk: Ping? While you're at it, might as well clean this up...

Comment (godbyk) [inline .travis.yml]: @tkoeppe Whoops! I missed that one. I'll fix it. Thanks!

**Comment by tkoeppe** [comment]
@godbyk: Are you sure this testing catches missing xref and grammar regen? I just pushed a branch that was missing the regeneration, and all four tests passed.

**Comment by godbyk** [comment]
@tkoeppe What do you mean by 'catches missing xref and grammar regen'?

It just runs `git status` on the generated files to see if they've changed from what's in the repository. The output of that command appears in the console output, but it doesn't throw a build error or anything.

(We should probably write better tests that do throw errors.)

**Comment by tkoeppe** [comment]
@godbyk Oh - but who reads the 5MB of console output? I thought it should be a test failure if you end up affecting the xrefs without checking them in yourself. Is that possible?

**Comment by godbyk** [comment]
@tkoeppe I think it should be possible. Lemme take a look at the Travis CI docs and get back to you.

(The `dot`-generated figures seem to always differ from the files in the git repository, but I think that's just due to the PDF metadata. We should find a better test for those.)

Are there any other tests we should run?

**Comment by godbyk** [comment]
@tkoeppe See https://github.com/cplusplus/draft/pull/780.
